src/Microsoft.IdentityModel.Protocols/Configuration/ConfigurationManager_Blocking.cs

```diff
@@ -0,0 +1,163 @@
+                                new InvalidOperationException(
+                                    LogHelper.FormatInvariant(LogMessages.IDX20803, LogHelper.MarkAsNonPII(MetadataAddress ?? "null"), LogHelper.MarkAsNonPII(_syncAfter), LogHelper.MarkAsNonPII(ex)), ex));
+                        }
+                        else
+                        {
+                            _syncAfter = DateTimeUtil.Add(
+                                TimeProvider.GetUtcNow().UtcDateTime,
+                                AutomaticRefreshInterval < RefreshInterval ? AutomaticRefreshInterval : RefreshInterval);
+
+                            var elapsedTime = TimeProvider.GetElapsedTime(startTimestamp);
+
+                            TelemetryClient.LogConfigurationRetrievalDuration(
+                                MetadataAddress,
+                                elapsedTime,
+                                ex);
+
+                            LogHelper.LogExceptionMessage(
+                                new InvalidOperationException(
+                                    LogHelper.FormatInvariant(LogMessages.IDX20806, LogHelper.MarkAsNonPII(MetadataAddress ?? "null"), LogHelper.MarkAsNonPII(ex)), ex));
+                        }
+                    }
+                }
+
+                // Stale metadata is better than no metadata
+                if (_currentConfiguration != null)
+                    return _currentConfiguration;
+                else
+                    throw LogHelper.LogExceptionMessage(
+                        new InvalidOperationException(
+                            LogHelper.FormatInvariant(
+                                LogMessages.IDX20803,
+                                LogHelper.MarkAsNonPII(MetadataAddress ?? "null"),
+                                LogHelper.MarkAsNonPII(_syncAfter),
+                                LogHelper.MarkAsNonPII(_fetchMetadataFailure)),
+                            _fetchMetadataFailure));
+            }
+            finally
+            {
+                _refreshLock.Release();
+            }
+        }
+
+        private void RequestRefreshBlocking()
+        {
+            DateTime now = TimeProvider.GetUtcNow().UtcDateTime;
+
+            if (now >= DateTimeUtil.Add(_lastRequestRefresh.UtcDateTime, RefreshInterval) || _isFirstRefreshRequest)
+            {
+                _refreshRequested = true;
+                _syncAfter = now;
+                _isFirstRefreshRequest = false;
+            }
+        }
+
+        private void TelemetryForUpdateBlocking()
+        {
+            string updateMode;
+
+            if (_currentConfiguration is null)
+                updateMode = TelemetryConstants.Protocols.FirstRefresh;
+            else
+                updateMode = _refreshRequested ? TelemetryConstants.Protocols.Manual : TelemetryConstants.Protocols.Automatic;
+
+            try
+            {
+                TelemetryClient.IncrementConfigurationRefreshRequestCounter(
```

Comment:
How would we differentiate between blocking and non-blocking calls in telemetry?

Reply:
Good question, I read the app context switch in the telemetry client and write that as a tag